docs/content/TabPanels.mdx

```diff
@@ -0,0 +1,30 @@
+---
+componentId: tab_panels
+title: TabPanels
+status: Alpha
+source: https://github.com/primer/react/blob/main/src/TabPanels
+---
+
+import data from '../../packages/react/src/TabPanels/TabPanels.docs.json'
+import DeprecationBanner from '../components/DeprecationBanner'
+
+<DeprecationBanner replacementUrl={'/components/tab-panels/react/latest'} />
+
+**Attention:** Make sure to properly label your `TabPanels` with an `aria-label` to provide context about the subject of your `TabPanels`.
+
+## Examples
+
+```jsx live
+<TabPanels aria-label="Select a tab">
+  <TabPanels.Tab>Tab 1</TabPanels.Tab>
+  <TabPanels.Tab>Tab 2</TabPanels.Tab>
+  <TabPanels.Tab>Tab 3</TabPanels.Tab>
+  <TabPanels.Panel>Panel 1</TabPanels.Panel>
+  <TabPanels.Panel>Panel 2</TabPanels.Panel>
+  <TabPanels.Panel>Panel 3</TabPanels.Panel>
+</TabPanels>
```

Comment:
Suggestion: if every `Panel` should always have a `Tab` and vice versa (a 1:1 relationship) could we combine the two into a single component? This would enforce safe usage ensuring the 1:1 relationship is always maintained and things are properly linked.

With this API it's pretty easy to accidentally break the 1:1 relationship & linking (as it is here actually - the tabs are not linked to the panels by `aria-labelledby` in this example).

In other words, the API would look something like this:

```suggestion
<TabPanels aria-label="Select a tab">
  <TabPanels.TabPanel name="Tab 1">
    Panel 1
  </TablPanels.TabPanel>
  <TabPanels.TabPanel name="Tab 2">
    Panel 2
  </TablPanels.TabPanel>
  <TabPanels.TabPanel name="Tab 3">
    Panel 3
  </TablPanels.TabPanel>
</TabPanels>
```

Under the hood it should be pretty straightforward to have `TabPanel` render `Tab` and `Panel`:

```tsx
const TabPanel = ({name, children}) => {
  const id = useId()
  
  return <>
    <TabPanel.Tab id={id}>{name}</TabPanel.Tab>
    <TabPanel.Panel aria-labelledby={id}>{children}</TabPanel.Tab>
  </>
}
```

Although if the order of the elements is important (panels must follow tabs) it would get a bit more complicated. But even then I think it would still be worth it to have a cleaner API.

Reply:
We could offer this API just within React quite easily but as @owenniblock points out we lose out on the ability to easily add content before or after the tabs, without doing a bunch of juggling of props and I don't know how prevalent it is to stick JSX in props but it doesn't make for a great API (IMO stuffing JSX in a prop would be an anti-pattern). In other words for very simple cases (as described above) it looks like a more desirable API but as we start to do the things we need the JSX devolves into a bit of a mess:

```jsx
<TabPanels aria-label="Select a tab">
  {/* each one of these could have icons or other markup that's kind of janky as a prop */}
  <TabPanels.TabPanel name={<><OctocatIcon /> Tab 1</>}>
    Panel 1
  </TablPanels.TabPanel>
  <TabPanels.TabPanel name={<><OctocatIcon /> Tab 2</>}>
    Panel 2
  </TablPanels.TabPanel>
  {/* I guess the best way to handle the Before/After tabs stuff would be another sub-component */}
  <TabPanels.AfterTabs>
    <MarkdownToolbar/>
  </TabPanels.AfterTabs>
  {/* Then because _eveything_ is out of tree order, anything after tab panels also needs a sub-component */}
  <TabPanel.AfterPanels>
    <ExtraButtons/>
  </TabPanel.AfterPanels>
</TabPanels>
```

vs 

```jsx
<TabPanels aria-label="Select a  tab">
  {/* Tabs don't need JSX props, can just render children */}
  <TabPanels.Tab><OctocatIcon /> Tab 1</TabPanels.Tab>
  <TabPanels.Tab><OctocatIcon /> Tab 2</TabPanels.Tab>
  {/* Stuff after tabs goes after the tabs in the tree */}
  <MarkdownToolbar/>
  <TablPanels.Panel>
    Panel 1
  </TablPanels.Panel>
  <TabPanels.Panel>
    Panel 2
  </TablPanels.Panel>
  {/* Stuff after panels goes after the panels in the tree */}
  <ExtraButtons/>
</TabPanels>
```

Reply:
> we're aiming for feature parity with Primer View Components here, so any API change we add here will need to be ported over to PVC in future

I think this is an unfortunate approach because ViewComponents and React are different paradigms. ViewComponents are consumed in HTML, which is not typechecked and doesn't have embedded JS. React components are consumed in typechecked JavaScript. This naturally means the APIs should be different - writing a React API to be the same a ViewComponent API is going to result a worse developer experience.

> IMO stuffing JSX in a prop would be an anti-pattern

FWIW this is a common practice and I personally don't think it's problematic at all, though admittedly it's not beautiful. This was actually recommended by the official React docs until very recently (it got refactorted out as they built the new site). It's flexible and allows for accepting strings for the basic case while still accepting full JSX for advanced cases.

In practice, the example with the current API really looks like this:

```tsx
const tab1Id = useId()
const tab2Id = useId()

<TabPanels aria-label="Select a  tab">
  <TabPanels.Tab id={tab1Id}><OctocatIcon /> Tab 1</TabPanels.Tab>
  <TabPanels.Tab id={tab2Id}><OctocatIcon /> Tab 2</TabPanels.Tab>
  <MarkdownToolbar/>
  <TablPanels.Panel aria-labelledby={tab1Id}>
    Panel 1
  </TabPanels.Panel>
  <TabPanels.Panel aria-labelledby={tab2Id}>
    Panel 2
  </TabPanels.Panel>
  <ExtraButtons/>
</TabPanels>
```

While it is flexible, it's problematic:

- It's difficult to remember where the IDs go (tabs or panels?)
- It's easy to mix up the IDs and link the wrong tabs with panels
- There's no clear link between the tabs & panels - as a reader who hasn't seen this before, would you intuitively understand that the tabs control the panels?

If the current level of flexibility is a requirement, I have a few suggestions that I think will improve the current API:

1. Make the `id` and `aria-labelledby` props required in the TypeScript types so they cannot be left out accidentally
2. Rename the `aria-labelledby` prop to `tabId`
3. Throw a runtime error if there is not a valid 1:1 link between tabs and panels, so we can catch these issues early
4. (This one is a little more complex) Create a new hook that returns tab and panel components already linked together, generating a unique ID under the hood

In practice (3) might look like this:

```tsx
const [Tab1, TabPanel1] = useTabPanel()
const [Tab2, TabPanel2] = useTabPanel()

<TabPanels aria-label="Select a  tab">
  <Tab1><OctocatIcon /> Tab 1</Tab1>
  <Tab2><OctocatIcon /> Tab 2</Tab2>
  <MarkdownToolbar/>
  <TabPanel1>
    Panel 1
  </TabPanel1>
  <TabPanel2>
    Panel 2
  </TabPanel2>
  <ExtraButtons/>
</TabPanels>
```

Alternatively if we prefer namespacing it could look like this:

```tsx
const Tab1 = TabPanels.useTab()
const Tab2 = TabPanels.useTab()

<TabPanels aria-label="Select a  tab">
  <Tab1.Tab><OctocatIcon /> Tab 1</Tab1.Tab>
  <Tab2.Tab><OctocatIcon /> Tab 2</Tab2.Tab>
  <MarkdownToolbar/>
  <Tab1.Panel>
    Panel 1
  </Tab1.Panel>
  <Tab2.Panel>
    Panel 2
  </Tab2.Panel>
  <ExtraButtons/>
</TabPanels>
```


I think this is more readable, clearly shows the link between the tabs and panels, and has fewer pitfalls. It means the consumer doesn't have to think about creating an ID, ensuring that it's unique, and knowing where to put it.

Reply:
@iansan5653 thanks for the valuable thoughts, I'm working on a change so that the `TabPanels` element requires an `id` and then that `id` is used to auto generate ids for the tabs, and wire up the `aria-labelledby` as well.

I'll let you know when that's pushed up so you can check it out and let me know if you think that addresses some of this feedback.